### PR TITLE
[Hardware][Ascend] Add silu_and_mul / rope ops and replace ops in attention layer

### PR DESCRIPTION
**Pull request by whx-sjtu**
add two custom_ops: silu_and_mul and rope
replace three ops in attention layer: reshape_and_cache, pagedattention and flashattention
change the profile run strategy

### CODE REVIEW

File: vllm_ascend/model_runner.py [hunk 1]
```diff
@@ -523,22 +545,16 @@ def profile_run(self) -> None:
...
-        # multiplying the list, to avoid Dynamo from treating them as
-        # tensor aliasing.
-        kv_caches = [
-            torch.tensor([], dtype=torch.float32, device=self.device)
-            for _ in range(num_layers)
-        ]
         finished_requests_ids = [seq.request_id for seq in seqs]
         model_input = self.prepare_model_input(
             seqs, finished_requests_ids=finished_requests_ids)
+        kv_caches, block_tables, slots = self.create_dummy_kv_cache(model_input.attn_metadata, model_input.input_tokens)
```
Comment (wangxiyuan) [inline vllm_ascend/model_runner.py:554]: can you merge `create_dummy_kv_cache` into `prepare_model_input` to generate model_input in one step? And if it is only used inner the class, raname to `_create_dummy_kv_cache`

File: vllm_ascend/attention.py [hunk 2]
```diff
@@ -25,6 +27,60 @@
...
+        mask_value = 1
+    attn_mask = torch.masked_fill(
+        torch.zeros(size=(max_seq_len, max_seq_len)), mask_flag, mask_value
+    ).to(dtype)
+    return attn_mask
+
+
+class AttentionMaskBuilder:
+    def __init__(self, attn_mask: torch.Tensor):
+        super().__init__()
```
Comment (wangxiyuan) [inline vllm_ascend/attention.py:50]: There is no base class for AttentionMaskBuilder, so super is useless here IMO

File: vllm_ascend/attention.py [hunk 3]
```diff
@@ -298,43 +324,30 @@ def decode_metadata(self) -> Optional["AscendMetadata"]:
...
-        range_end = seq_len
-        numel = range_end - range_start
-        block_table = block_tables[seq_id]
-
-        for i in range(range_start, range_end):
-            block_number = block_table[i // block_size]
-            block_offset = i % block_size
-            slot_indices.append([block_number, block_offset])
-        slot_indices.extend([[PAD_SLOT_ID, 0]] * (max_query_len - numel))
+    attn_mask_builder = None
```
Comment (wangxiyuan) [inline vllm_ascend/attention.py:327]: any reason tht make `attn_mask_builder` a class property? Why not initialize it in `__init__` method?

**Review by wangxiyuan — COMMENTED** [review]
Please rebase after the CI check in enabled. There are many lint errors need fix. @MengqingCao  please enable the CI for all branch ASAP. Thanks

File: vllm_ascend/model_runner.py [hunk 4]
```diff
@@ -105,41 +106,16 @@ def build(self) -> ModelInputForGPU:
...
-                    0,
-                    dtype=torch.int,
-                    device=self.runner.device)
-                input_positions_tensor = torch.flatten(input_positions_tensor)
-
-            max_seq_len = max(seq_lens)
-            seq_lens = len(seq_lens) * [max_seq_len]
-        else:
-            input_tokens_tensor = torch.tensor(flatten_2d_lists(input_tokens),
+        input_tokens_tensor = torch.tensor(flatten_2d_lists(input_tokens),
```
Comment (MengqingCao) [inline vllm_ascend/model_runner.py:109]: QQ: Is the attention op in torch-npu support ver-len input now?

File: vllm_ascend/attention.py (hunk 3 above)
Comment (whx-sjtu) [reply vllm_ascend/attention.py:327]: because this AscendMetadataBuilder class might be instantiatied many times, but we only want to construct the attenion mask once for performance consideration.

File: vllm_ascend/attention.py (hunk 2 above)
Comment (whx-sjtu) [reply vllm_ascend/attention.py:50]: thanks, i will fix this

File: vllm_ascend/model_runner.py (hunk 1 above)
Comment (whx-sjtu) [reply vllm_ascend/model_runner.py:554]: I make this function defined alone because it is only used in warmup stage while prepare_model_input is used for normal execute_model. Maybe it can be merged into profile run, but my concern is that this might make profile_run too long. I will rename this function first.

File: vllm_ascend/model_runner.py (hunk 4 above)
Comment (whx-sjtu) [reply vllm_ascend/model_runner.py:109]: yes, the attention op we use supports var_len input, so padding is not needed here

File: vllm_ascend/attention.py (hunk 3 above)
Comment (wangxiyuan) [reply vllm_ascend/attention.py:327]: So, every time `AscendMetadataBuilder` initialized , it use the same `AttentionMaskBuilder.initialize_from_len(128, self.input_builder.runner.model_config.dtype)`?  `self.input_builder` is always the same？

File: vllm_ascend/model_runner.py (hunk 1 above)
Comment (wangxiyuan) [reply vllm_ascend/model_runner.py:554]: I didn't see `prepare_model_input ` or `create_dummy_kv_cache` is called in any other place. Correct me If I missed something. Or It'll be done in the future PRs?

Comment (wangxiyuan) [reply vllm_ascend/model_runner.py:554]: Got, just realize that `prepare_model_input`  is used in some other scenario

File: vllm_ascend/attention.py (hunk 3 above)
Comment (whx-sjtu) [reply vllm_ascend/attention.py:327]: `AttentionMaskBuilder.initialize_from_len` should only be called once globally in normal situation, in the first initialization of `AscendMetadataBuilder`

File: vllm_ascend/attention.py [hunk 5]
```diff
@@ -438,12 +533,15 @@ def __init__(
         assert self.num_heads % self.num_kv_heads == 0
         self.num_queries_per_kv = self.num_heads // self.num_kv_heads
 
+        # used for prefill calculation
+        self.prefill_blocktable = torch.tensor([0], dtype=torch.int32, device="npu")
```
Comment (MengqingCao) [inline vllm_ascend/attention.py:537]: Seems it is unused, or I missed something?

File: vllm_ascend/attention.py [hunk 6]
```diff
@@ -570,88 +645,21 @@ def forward(
...
             # FA for decoding phase
             assert kv_cache is not None
-            # shape of query [B,S*H] --> [B,S,H]
-            query = query.view(
-                -1,
-                1,
-                self.head_size * self.num_heads,
-            )
-            output = torch_npu.npu_incre_flash_attention(
+            output = torch.empty(num_tokens, self.num_heads, self.head_size, dtype=query.dtype, device="npu")
```
Comment (MengqingCao) [inline vllm_ascend/attention.py:648]: `num_tokens == 1` will always be true while decoding, could we just use a `self.output` or a global npu tensor to avoid allocate mem on npu frequently?

File: vllm_ascend/attention.py [hunk 7]
```diff
@@ -25,6 +27,61 @@
...
 
+def generate_attn_mask(max_seq_len: int, dtype=torch.float16):
+    # lower triangle matrix
+    mask_flag = torch.tril(
+        torch.ones((max_seq_len, max_seq_len),
+                   dtype=torch.bool)).view(max_seq_len, max_seq_len)
+    # upper triangle matrix used to mark mask positions
+    mask_flag = ~mask_flag
+    # for fp16 dtype, the mask value should be set to -inf
+    if dtype == torch.float16:
```
Comment (wuhuikx) [inline vllm_ascend/attention.py:38]: how about the other datatype? BF16、FP32 and INT8?

File: vllm_ascend/attention.py [hunk 8]
```diff
@@ -141,12 +184,8 @@ class AscendMetadata(AttentionMetadata, PagedAttentionMetadata):
...
     # the computed tokens + new tokens None if it is a decoding.
     seq_lens: Optional[List[int]] = None
-
-    # FIXME: It is for flash attn.
-    # (batch_size + 1,). The cumulative sequence lengths of the sequences in
-    # the batch, used to index into sequence. E.g., if the sequence length is
-    # [4, 6], it is [0, 4, 10].
-    seq_start_loc: Optional[torch.Tensor] = None
+    # seq_lens stored as a cpu tensor.
+    seq_lens_tensor_cpu: Optional[torch.Tensor] = None
```
Comment (wuhuikx) [inline vllm_ascend/attention.py:188]: Does this means, we need a “tensor” instead of List[int] to present seq_lens info?

File: vllm_ascend/attention.py [hunk 9]
```diff
@@ -399,12 +407,101 @@ def _add_seq_group(
...
+            block_tables = make_tensor_with_pad(
+                self.block_tables,
+                pad=0,
+                dtype=torch.int32,
+                device=device,
+            )
+        assert max_query_len > 0, "query_lens: {}".format(query_lens)
+
+        assert device is not None
+        context_lens_tensor = async_tensor_h2d(self.context_lens, torch.int32,
```
Comment (wuhuikx) [inline vllm_ascend/attention.py:475]: this H2D will happen on each decode iteration, right?

File: vllm_ascend/attention.py [hunk 10]
```diff
@@ -399,12 +407,101 @@ def _add_seq_group(
...
+                                               device, self.runner.pin_memory)
+        seq_lens_tensor = async_tensor_h2d(seq_lens, torch.int32, device,
+                                           self.runner.pin_memory)
+        seq_lens_tensor_cpu = torch.tensor(seq_lens, dtype=torch.int32)
```
Comment (wuhuikx) [inline vllm_ascend/attention.py:479]: why we need both seq_lens_tensor and seq_lens_tensor_cpu and the same time?

File: vllm_ascend/attention.py [hunk 11]
```diff
@@ -474,184 +571,54 @@ def forward(
...
-                    input_layout="BNSD",
-                    num_key_value_heads=self.num_kv_heads,
-                    pre_tokens=65535,
-                    next_tokens=0,
-                    sparse_mode=attn_metadata.sparse_mode,
-                )
-                # reshape to [B,H]
-                output = output.transpose(1, 2).reshape(
-                    num_tokens, self.num_heads * self.head_size)
+                torch_npu.npu_selfattention(query, key, value, mask,
```
Comment (wuhuikx) [inline vllm_ascend/attention.py:606]: actually, there are several FA implementation in CANN library, do we need a dispatch strategy here to choose different kernels for different situations ?

File: vllm_ascend/attention.py [hunk 12]
```diff
@@ -474,184 +571,54 @@ def forward(
...
-                        block_table=attn_metadata.block_tables,
-                        block_size=key_cache.
-                        shape[1],  # max val of block_size == 512
-                        actual_seq_lengths=attn_metadata.seq_lens,
-                    )
-                # [B,S,H] --> [B,H]
-                output = output.squeeze(1)
-
+                raise RuntimeError(
+                    "Prefix cache and chunked prefill is currently not supported."
```
Comment (wuhuikx) [inline vllm_ascend/attention.py:612]: why not support prefix cache and chunked prefill?

File: vllm_ascend/ops/rotary_embedding.py [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+    offsets: Optional[torch.Tensor] = None,
+) -> Tuple[torch.Tensor, torch.Tensor]:
+    import torch_npu
+
+    self.cos_sin_cache = self.cos_sin_cache.to(query.device, dtype=query.dtype)
+    if offsets is not None:
+        raise NotImplementedError(
+            "Batched rotary embedding is currently not supported on NPU.")
+    else:
+        query = query.contiguous()
```
Comment (wuhuikx) [inline vllm_ascend/ops/rotary_embedding.py:21]: why we need contiguous here? Actually, the operator of “npu_rope” should check the tensor contiguous.

File: vllm_ascend/worker.py [hunk 14]
```diff
@@ -131,6 +131,12 @@ def determine_num_available_blocks(self) -> Tuple[int, int]:
         # Execute a forward pass with dummy inputs to profile the memory usage
         # of the model.
         self.model_runner.profile_run()
+        # get the num of kv blocks used in warmup stage
```
Comment (wuhuikx) [inline vllm_ascend/worker.py:134]: please aware that we need a full sentence for comment and don't forget to add the “dot”.

File: vllm_ascend/attention.py (hunk 7 above)
Comment (whx-sjtu) [reply vllm_ascend/attention.py:38]: for other datatype the mask value is 1, only for fp16 datatype it should be set to min of float

File: vllm_ascend/attention.py (hunk 9 above)
Comment (whx-sjtu) [reply vllm_ascend/attention.py:475]: yes it is, so here uses non-blocking H2D transfer

File: vllm_ascend/worker.py (hunk 14 above)
Comment (whx-sjtu) [reply vllm_ascend/worker.py:134]: got it

File: vllm_ascend/worker.py [hunk 15]
```diff
@@ -131,6 +131,12 @@ def determine_num_available_blocks(self) -> Tuple[int, int]:
         # Execute a forward pass with dummy inputs to profile the memory usage
         # of the model.
         self.model_runner.profile_run()
+        # Get the number of kv blocks used in warmup stage.
+        dummy_block_num = self.model_runner.dummy_block_num if getattr(self.model_runner, "dummy_block_num", None) is not None \
+            else self.model_runner._base_model_runner.dummy_block_num
+        block_size = self.cache_config.block_size
+        dummy_block_size = 128
+        dummy_num_blocks = dummy_block_size // block_size * dummy_block_num
```
Comment (ganyi1996ppo) [inline vllm_ascend/worker.py:139]: Why do we need this?

File: vllm_ascend/attention.py (hunk 7 above)
Comment (ganyi1996ppo) [reply vllm_ascend/attention.py:35]: This sound strange, can you add a todo on this func? I believe we should eliminate this part in the future.

File: vllm_ascend/worker.py (hunk 15 above)
Comment (whx-sjtu) [reply vllm_ascend/worker.py:139]: for more accurate NPU memory calculation during warmup

File: vllm_ascend/attention.py [hunk 16]
```diff
@@ -399,12 +399,101 @@ def _add_seq_group(
...
+
+        if self.num_prefills > 0:
+            self.attn_mask = AscendMetadataBuilder._attn_mask_builder.get_attn_mask(  # type: ignore
+                max_prefill_seq_len,
+                self.input_builder.runner.model_config.dtype,
+                self.input_builder.runner.device)
+        else:
+            self.attn_mask = None
+
+        if use_captured_graph:
```
Comment (ganyi1996ppo) [inline vllm_ascend/attention.py:444]: Do we have any graph needs to be  captured? BTW, we should remove any cuda related keywords which makes this PR kinds of weird........

Comment (whx-sjtu) [reply vllm_ascend/attention.py:444]: This comes from vllm-ascend main repo, I haven't fix these codes yet. : ) Will remove all of these in a dedicated PR.

Comment (wuhuikx) [reply vllm_ascend/attention.py:444]: same as ganyi, please remove

File: vllm_ascend/attention.py [hunk 17]
```diff
@@ -399,12 +399,101 @@ def _add_seq_group(
...
+            self.slot_mapping.extend([PAD_SLOT_ID] * cuda_graph_pad_size)
+            self.block_tables.extend([] * cuda_graph_pad_size)
```
[the remark this answers is not in the document]
Comment (wuhuikx) [reply vllm_ascend/attention.py:446]: same as ganyi, please remove

**Review by wangxiyuan — APPROVED** [review]
I'm fine with the PR. Please address the comments from  @wuhuikx and @ganyi1996ppo  and make the CI happy.

File: vllm_ascend/ops/rotary_embedding.py (hunk 13 above)
Comment (whx-sjtu) [reply vllm_ascend/ops/rotary_embedding.py:21]: will remove these when new op is ready

File: vllm_ascend/attention.py (hunk 12 above)
Comment (whx-sjtu) [reply vllm_ascend/attention.py:576]: because param q_len is currently not supportted in pagedattention of torch_npu. working on it.

File: vllm_ascend/attention.py (hunk 7 above)
Comment (whx-sjtu) [reply vllm_ascend/attention.py:35]: got it

File: vllm_ascend/attention.py (hunk 8 above)
Comment (whx-sjtu) [reply vllm_ascend/attention.py:175]: yes, this is currently needed by our op

File: vllm_ascend/attention.py (hunk 10 above)
Comment (whx-sjtu) [reply vllm_ascend/attention.py:445]: device tensor `seq_lens_tensor` is needed for broadcast, while host tensor `seq_lens_tensor_cpu` is needed by pa/fa op :). will try to eliminate this in the future.

File: vllm_ascend/attention.py (hunk 11 above)
Comment (whx-sjtu) [reply vllm_ascend/attention.py:570]: yes indeed. but currenly we only support one FA kernel.

File: vllm_ascend/attention.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (whx-sjtu) [reply vllm_ascend/attention.py]: removed

File: vllm_ascend/attention.py (hunk 16 above)
Comment (whx-sjtu) [reply vllm_ascend/attention.py]: removed :)

File: vllm_ascend/attention.py (hunk 5 above)
Comment (whx-sjtu) [reply vllm_ascend/attention.py]: removed

File: vllm_ascend/attention.py (hunk 6 above)
Comment (whx-sjtu) [reply vllm_ascend/attention.py]: in decoding phase `num_tokens` is equal to batch size which can be changed

File: vllm_ascend/__init__.py [hunk 18]
```diff
@@ -1,3 +1,15 @@
 def register():
     """Register the NPU platform."""
+    register_mindie_turbo()
     return "vllm_ascend.platform.NPUPlatform"
+
+
+# flake8: noqa: F401
+def register_mindie_turbo():
+    try:
+        from mindie_turbo import vllm_turbo  # type: ignore
```
Comment (wuhuikx) [inline vllm_ascend/__init__.py:10]: Can we add a logger info here:

"MindIE Turbo is installed. LLM inference is speedup with MindIE Turbo."

File: vllm_ascend/__init__.py [hunk 19]
```diff
@@ -1,3 +1,15 @@
...
+    except ImportError:
+        from vllm.logger import init_logger
+        logger = init_logger(__name__)
+        logger.info(
```
Comment (wuhuikx) [inline vllm_ascend/__init__.py:14]: remove this log, because it is default to use ascend-vllm without turbo.

File: vllm_ascend/__init__.py (hunk 18 above)
Comment (whx-sjtu) [reply vllm_ascend/__init__.py:10]: got it.

Comment (wuhuikx) [reply vllm_ascend/__init__.py:10]: "MindIE Turbo is installed. vLLM inference efficiency will be accelerated with MindIE Turbo."  this is much more clear. Can you help to change?

File: vllm_ascend/__init__.py (hunk 19 above)
Comment (whx-sjtu) [reply vllm_ascend/__init__.py:16]: done.

File: vllm_ascend/__init__.py (hunk 18 above)
Comment (whx-sjtu) [reply vllm_ascend/__init__.py]: done!

**Review by wuhuikx — APPROVED** [review]

**Review by ganyi1996ppo — APPROVED** [review]
